### PR TITLE
Put session ID in ALPN

### PR DESCRIPTION
**Pull request by jefferai**
This allows us to use SNI for actual host (e.g. routing) information. We
don't have the client cert yet so we can't look at that, so this
provides a way for us to convey the information needed to look up that
session's TLS stack.

Using SNI for hosts means we also run into the fact that we don't have
automatic agreement in terms of SANs. So when generating the certs we
now also pass worker address information to the function to be encoded
in the cert.

Finally, there is a change in how the websocket dialing happens, because
http.RoundTripper tries to be too clever for its own good and overwrites
NextProtos on a whim.

### CODE REVIEW

File: internal/session/repository_session.go [hunk 1]
```diff
@@ -64,12 +64,14 @@ func (r *Repository) CreateSession(ctx context.Context, sessionWrapper wrapping.
...
 
+	opts := getOpts(opt...)
+
 	id, err := newId()
 	if err != nil {
 		return nil, nil, errors.Wrap(ctx, err, op)
 	}
 
-	privKey, certBytes, err := newCert(ctx, sessionWrapper, newSession.UserId, id, newSession.ExpirationTime.Timestamp.AsTime())
+	privKey, certBytes, err := newCert(ctx, sessionWrapper, newSession.UserId, id, opts.withWorkerAddresses, newSession.ExpirationTime.Timestamp.AsTime())
```
Comment (jimlambrt) [inline internal/session/repository_session.go:74]: Is this really optional?  Feels like a required parameter.

File: internal/session/session.go [hunk 2]
```diff
@@ -351,7 +352,8 @@ func contains(ss []string, t string) bool {
 	return false
 }
 
-func newCert(ctx context.Context, wrapper wrapping.Wrapper, userId, jobId string, exp time.Time) (ed25519.PrivateKey, []byte, error) {
+// newCert creates a new session certificate. If addresses are supplied, they will be parsed and added to IP or DNS SANs as appropriate.
```
Comment (jimlambrt) [inline internal/session/session.go:355]: should we allow addresses to be empty?

File: internal/session/repository_session.go [hunk 3]
```diff
@@ -19,7 +19,7 @@ import (
 // its State of "Pending".  The following fields must be empty when creating a
 // session: ServerId, ServerType, and PublicId.  No options are
 // currently supported.
-func (r *Repository) CreateSession(ctx context.Context, sessionWrapper wrapping.Wrapper, newSession *Session, _ ...Option) (*Session, ed25519.PrivateKey, error) {
+func (r *Repository) CreateSession(ctx context.Context, sessionWrapper wrapping.Wrapper, newSession *Session, opt ...Option) (*Session, ed25519.PrivateKey, error) {
```
Comment (jimlambrt) [inline internal/session/repository_session.go:22]: should addrs be optional?  if so, then mind changing the godoc?

Comment (jimlambrt) [reply internal/session/repository_session.go:22]: also, might be nice to have unit tests that require addresses.

File: internal/servers/worker/handler.go [hunk 4]
```diff
@@ -52,11 +53,26 @@ func (w *Worker) handleProxy(listenerCfg *listenerutil.ListenerConfig) (http.Han
 	return func(wr http.ResponseWriter, r *http.Request) {
```
Comment (jimlambrt) [inline internal/servers/worker/handler.go:53]: I know writing handler tests is a pain, but should we add at least one that verifies the error handling?

File: internal/servers/worker/worker.go [hunk 5]
```diff
@@ -265,11 +266,20 @@ func (w *Worker) getSessionTls(hello *tls.ClientHelloInfo) (*tls.Config, error)
 	ctx := w.baseContext
```
Comment (jimlambrt) [inline internal/servers/worker/worker.go:266]: might be nice to have a test for the unexported func, given how buried it is during runtime...

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Mostly questions/suggestions about tests.  With that said, I do wonder if addrs should be an option for creating a session?

**Comment by jimlambrt** [comment]
Mind updating go.mod with: go get -u golang.org/x/sys

see: https://stackoverflow.com/questions/71507321/go-1-18-build-error-on-mac-unix-syscall-darwin-1-13-go253-golinkname-mus

File: internal/session/repository_session.go (hunk 1 above)
Comment (jefferai) [reply internal/session/repository_session.go:74]: Yes, it's optional. I just didn't want to introduce an options pattern into `newCert` -- but I can easily do so if you prefer!

File: internal/session/session.go (hunk 2 above)
Comment (jefferai) [reply internal/session/session.go:355]: I don't think it hurts anything. It's not always necessary. In practice we'll probably always supply something except in test code.

If you prefer I could ensure it's not empty and update tests, but I don't have a preference.

**Review by jimlambrt — APPROVED** [review]
ty!

**Pull request merged**